specs/waku/waku-v2.md

```diff
@@ -216,37 +216,103 @@ The `uuid` field MUST indicate which query is being responded to.
+Content filtering is a way to do [message-based
 filtering](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern#Message_filtering).
 Currently the only content filter being applied is on `contentTopic`. This
 corresponds to topics in Waku v1.
 
-A node that only sets this field but doesn't subscribe to any topic SHOULD only
-get notified when the content subtopic matches. A content subtopic matches when
-a `Wakumessage` `contentTopic` field is the same.
+### Rationale
+
+Unlike the `store` protocol for historical messages, this protocol allows for
+native lower latency scenarios such as instant messaging. It is thus
+complementary to it.
+
+Strictly speaking, it is not just doing basic request response, but performs
+sender push based on receiver intent. While this can be seen as a form of light
+pub/sub, it is only used between two nodes in a direct fashion. Unlike the
+Gossip domain, this is meant for light nodes which put a premium on bandwidth.
+No gossiping takes place.
 
-A node that receives this RPC SHOULD apply this content filter before relaying.
-Since such a node is doing extra work for a light node, it MAY also account for
-usage and be selective in how much service it provides. This mechanism is
-currently planned but underspecified.
+As a technical detail, the [Topic Validation used in libp2p
+PubSub](https://github.com/libp2p/specs/tree/master/pubsub#topic-validation)
+currently doesn't apply to individual nodes. This makes it unsuitable as a form
+of subscribe mechanism, in addition to peer management etc being expensive for
+light nodes.
+
+It is worth noting that a light node could get by with only using the `store`
+protocol to query for a recent time window, provided it is acceptable to do
+frequent polling.
 
 #### Protobuf
 
+TODO Consider adding a FilterResponse acting as a form of ACK
+
+TODO Consider adding request id
+
+TODO Clarify if Messages or a list of WakuMessage are pushed
+
+TODO Specify unsubscribe mechanism and semantics
+
 ```protobuf
 message RPC {
-  repeated ContentFilter contentFilter = 1;
+  repeated FilterRequest filterRequest = 1;
+  // space for optional FilterResponse
+  repeated MessagePush messagePush = 3;
 }
 
+message FilterRequest {
+  // space for optional request id
+  repeated ContentFilter contentFilter = 2;
+  optional string topic = 3;
+
   message ContentFilter {
     optional string contentTopic = 1;
   }
 }
+
+message MessagePush {
+  repeated Message message = 1;
+}
 ```
 
+##### FilterRequest
+
+A node that sends the RPC with a filter request requests that the filter node
+SHOULD notify the light requesting node of messages matching this filter.
+
+The filter matches when content filter and, optionally, a topic is matched.
+Content filter is matched when a `WakuMessage` `contentTopic` field is the same.
+
+A filter node SHOULD honor this request, though it MAY choose not to do so. If
+it chooses not to do so it SHOULD tell the light why. For notifying the light
```

Comment:
How should it tell this?

Reply:
Unspecified right now, I was thinking with a response/notification in future versions of the spec. E.g. something like:

1. Normal case: Filter Request -> Ok
2. No funds case: Filter Request -> NoFundError
3. Spam case: Filter Request -> NoFundError, FilterRequestFilterRequestFilterRequestFilterRequest -> **crickets/disconnect**

Reply:
Changed to MAY and added todo